encode/vp8enc.c

```diff
@@ -0,0 +1,1173 @@
+        current_surface = vaapi_context.surfaces[i];
+
+        if(current_surface != vaapi_context.last_ref_surface && current_surface != vaapi_context.golden_ref_surface && current_surface != vaapi_context.alt_ref_surface)
+          return current_surface;
+  }
+
+  //No unused surface found - should never happen.
+  fprintf(stderr, "Error: No unused surface found!\n");
+  assert(0);
+
+}
+
+VASurfaceID vp8enc_update_reference(VASurfaceID current_surface, VASurfaceID second_copy_surface, bool refresh_with_recon, int copy_flag)
+{
+  if(refresh_with_recon)
+    return vaapi_context.recon_surface;
+  switch(copy_flag)
+  {
+    case 0:
+      return current_surface;
+    case 1:
+      return vaapi_context.last_ref_surface;
+    case 2:
+      return second_copy_surface;
+    default: // should never happen
+      fprintf(stderr, "Error: Invalid copy_buffer_to_X flag\n");
+      assert(0);
+  }
+
+  return VA_INVALID_ID; // should never happen
+}
+
+
+void vp8enc_update_reference_list(int frame_type)
+{
+
+  VAEncPictureParameterBufferVP8 *picParam = &vaapi_context.pic_param;
+
+  if (frame_type == KEY_FRAME)
+  {
+     vaapi_context.last_ref_surface = vaapi_context.recon_surface;
+     vaapi_context.golden_ref_surface = vaapi_context.recon_surface;
+     vaapi_context.alt_ref_surface = vaapi_context.recon_surface;
+  } else { // INTER_FRAME
+     //check refresh_X and copy_buffer_to_golden_X and update references accordingly
+     if(picParam->pic_flags.bits.refresh_last)
+       vaapi_context.last_ref_surface = vaapi_context.recon_surface;
+     vaapi_context.golden_ref_surface = vp8enc_update_reference(vaapi_context.golden_ref_surface,vaapi_context.alt_ref_surface,picParam->pic_flags.bits.refresh_golden_frame, picParam->pic_flags.bits.copy_buffer_to_golden );
+     vaapi_context.alt_ref_surface = vp8enc_update_reference(vaapi_context.alt_ref_surface,vaapi_context.golden_ref_surface,picParam->pic_flags.bits.refresh_alternate_frame, picParam->pic_flags.bits.copy_buffer_to_alternate );
+  }
+
+  vaapi_context.recon_surface = vp8enc_get_unused_surface();
+}
+
+void vp8enc_init_MiscParameterBuffers(VAEncMiscParameterHRD *hrd, VAEncMiscParameterFrameRate *frame_rate, VAEncMiscParameterRateControl *rate_control)
+{
+  if(hrd != NULL)
+  {
+    if(settings.frame_bitrate)
+    {
+      hrd->initial_buffer_fullness = settings.frame_bitrate * settings.hrd_window / 2;
+      hrd->buffer_size = settings.frame_bitrate * settings.hrd_window;
+    } else {
+      hrd->initial_buffer_fullness = 0;
+      hrd->buffer_size = 0;
+    }
+  }
+
+  if(frame_rate != NULL)
+  {
+    frame_rate->framerate = settings.frame_rate;
+  }
+
+  if (rate_control != NULL)
+  {
+    rate_control->window_size = settings.hrd_window;
+    rate_control->initial_qp = settings.quantization_parameter;
+    rate_control->min_qp = settings.clamp_qindex_low;
+    //rate_control->rc_flags.bits.disable_bit_stuffing = 1;
+
+    if(settings.rc_mode == VA_RC_VBR)
```

Comment:
Is current driver support layered bitrate control? If yes we may need set bitrate for each layer.
for example, if we have 3 layers, 40% bit rate to layer0, 20% for layer 1, 40% for layer2....